src/main/java/net/logstash/logback/util/DelegatingContext.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2013-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.logstash.logback.util;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+
+import ch.qos.logback.core.Context;
+import ch.qos.logback.core.spi.LifeCycle;
+import ch.qos.logback.core.status.StatusManager;
+
+/**
+ * Logback {@link Context} implementation that delegate calls to another wrapped instance.
+ * 
+ * @author brenuart
+ */
+public class DelegatingContext implements Context {
```

Comment:
I'm a little worried about the Context interface evolving with new default methods, and us having to keep up with it.  (same for DelegatingStatusManager) Can you think of any other option that does not require these delegating objects

Reply:
Errors detected by the PatternLayout fall in two categories:

- **Syntax Error**. Happens when the parser fails to parse the pattern for reasons like a missing closing bracket. In this case, the PatternLayoutBase logs an ERROR status and returns an empty string when its `doLayout(event)` method is invoked afterwards. Technically, the converter chain is empty (the _head_ is null).

- **Conversion Error**. Happens when an unknown conversion word is used (`%name`) or when it fails to instantiate an instance of the corresponding converter. In this case, instead of throwing an exception, the Parser replaces the conversion word with the string `%PARSER_ERROR[name]`.
Suppose you want to refer to the logger level and use `%loggerLevel` instead of `%level`, then the PatternLayout will produce  `"%PARSE_ERROR[loggerLevel]"` at runtime. From a technical point of view, the conversion chain contains a single LiteralConverter with a constant value.

We could rely on this behaviour and inspect the conversion chain looking for LiteralConverters with   `%PARSE_ERROR[...]` in them. This strategy has the following limitations:
- syntax errors are not detected. We can deduce a syntax error occurred via an other mechanism but we won't be able to capture the actual cause - only the fact a syntax error occurred;
- we know a conversion error occurred but here again we don't have the cause (the cause is logged in the error status);

Having said that, this may not be a problem: the PatternLayoutBase will have logged the actual cause itself before we detect an error occurred. The StatusManager would report something like this:
```
11:15:32,589 |-ERROR in ch.qos.logback.core.pattern.parser.Compiler@4944252c - There is no conversion class registered for conversion word [loggerLevel]
11:15:32,590 |-ERROR in ch.qos.logback.core.pattern.parser.Compiler@4944252c - [loggerLevel] is not a valid conversion word
11:15:32,591 |-ERROR in net.logstash.logback.composite.loggingevent.LoggingEventPatternJsonProvider@2781e022 - Invalid [pattern]: Invalid JSON property '/level' (was '%loggerLevel'): failed to interpret '%loggerLevel' (see previous statuses for more information)
```

The current solution emits a single ERROR status with all the details:
```
11:35:36,541 |-ERROR in net.logstash.logback.composite.loggingevent.LoggingEventPatternJsonProvider@2781e022 - Invalid [pattern]: Invalid JSON property '/level' (was '%loggerLevel'): There is no conversion class registered for conversion word [loggerLevel]
```

Both approaches can do the job.
I preferred to capture errors by hooking into `StatusManager#addError()` instead of parsing strings looking for error patterns. This looked more reliable to me... This must be balanced against the maintenance overhead you mentionned.

Reply:
As far as implementation... Both approaches are quiet "ugly" (no offense to your code, it's just what we're forced to do given logback's behavior).  But I think the approach from gh686-patternlayout-errorhandling-alternative is more maintainable.

I think in either case, the user will easily realize something is wrong.

So, I'm leaning toward the gh686-patternlayout-errorhandling-alternative impl

Reply:
Ok. Sold.